internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -0,0 +1,154 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package variables
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	structuraldefaulting "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ValidateClusterClassVariables validates clusterClassVariable.
+func ValidateClusterClassVariables(clusterClassVariables []clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, validateClusterClassVariableNamesUnique(clusterClassVariables, fldPath)...)
+
+	for i := range clusterClassVariables {
+		allErrs = append(allErrs, validateClusterClassVariable(&clusterClassVariables[i], fldPath.Index(i))...)
+	}
+
+	return allErrs
+}
+
+// validateClusterClassVariableNamesUnique validates that ClusterClass variable names are unique.
+func validateClusterClassVariableNamesUnique(clusterClassVariables []clusterv1.ClusterClassVariable, pathPrefix *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	variableNames := sets.NewString()
+	for i, clusterClassVariable := range clusterClassVariables {
+		if variableNames.Has(clusterClassVariable.Name) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					pathPrefix.Index(i).Child("name"),
+					clusterClassVariable.Name,
+					fmt.Sprintf("variable names must be unique. %v is repeated.", clusterClassVariable.Name),
+				),
+			)
+		}
+		variableNames.Insert(clusterClassVariable.Name)
+	}
+
+	return allErrs
+}
+
+// validateClusterClassVariable validates a ClusterClassVariable.
+func validateClusterClassVariable(variable *clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Validate variable name.
+	allErrs = append(allErrs, validateClusterClassVariableName(variable.Name, fldPath.Child("name"))...)
+
+	// Validate schema.
+	allErrs = append(allErrs, validateSchema(&variable.Schema.OpenAPIV3Schema, variable.Name, fldPath.Child("schema", "openAPIV3Schema"))...)
+
+	return allErrs
+}
+
+// validateClusterClassVariableName validates a variable name.
+func validateClusterClassVariableName(variableName string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if variableName == "" {
+		allErrs = append(allErrs, field.Invalid(fldPath, variableName, "name cannot be empty"))
+	}
+	if strings.HasPrefix(variableName, "builtin") {
```

Comment:
```suggestion
	if variableName == "builtin" {
```
(+ corresponding change in the error message)

Sorry my bad. Let's switch back to equal here. I think it's good enough especially with the combinations of dots are not allowed

Reply:
Can you please also add a todo (you can assign it to me) that this should eventually probably be a shared constant with `controllers/topology/internal/extensions/patches/variables.BuiltinsName`